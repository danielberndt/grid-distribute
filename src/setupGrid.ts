import distribute, {UserOptions} from "./distribute";
import {ValidTiles} from "./types";

interface GridOptions {
  width: number;
  height: number;
}

interface OptionalGridOptions {
  maxRatio: number;
  minRatio: number;
}

type FullGridOptions = GridOptions & OptionalGridOptions;

const findValidTiles = ({width, height, minRatio, maxRatio}: FullGridOptions) => {
  const tiles: ValidTiles = {};
  const fullArea = width * height;
  for (let w = width; w >= 1; w -= 1) {
    for (let h = height; h >= 1; h -= 1) {
      const ratio = w / h;
      if (ratio >= minRatio && ratio <= maxRatio) {
        const dim = {width: w, height: h};
        const info = tiles[w * h];
        if (info) {
          info.tiles.push(dim);
        } else {
          tiles[w * h] = {areaRatio: (w * h) / fullArea, tiles: [dim]};
        }
      }
    }
  }
  return tiles;
};

const defaultGridOpts: OptionalGridOptions = {
  maxRatio: 2,
  minRatio: 0.5,
};

const setupGrid = (userOpts: GridOptions & Partial<OptionalGridOptions>) => {
  const opts = {
    ...defaultGridOpts,
    ...userOpts,
  };

  const tiles = findValidTiles(opts);

  return {
    tiles,
    distribute: <E>(distUserOpts: UserOptions<E>) =>
      distribute(distUserOpts, {tiles, width: opts.width, height: opts.height}),
  };
};

export default setupGrid;
