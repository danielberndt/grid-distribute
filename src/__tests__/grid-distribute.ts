import setupGrid from "../index";

describe("distribute", () => {
  it("works with one element", () => {
    const grid = setupGrid({width: 4, height: 3});
    const elements = [{prio: 2}];
    const result = grid.distribute({elements, getPriority: e => e.prio});
    expect(result).toMatchSnapshot();
  });

  it("works with two identical elements", () => {
    const grid = setupGrid({width: 4, height: 3});
    const elements = [{prio: 2}, {prio: 2}];
    const result = grid.distribute({elements, getPriority: e => e.prio});
    expect(result).toMatchSnapshot();
  });

  it("works with many elements", () => {
    const grid = setupGrid({width: 4, height: 3});
    const elements = [{prio: 2}, {prio: 2}, {prio: 1}, {prio: 4}, {prio: 0.5}];
    const result = grid.distribute({elements, getPriority: e => e.prio});
    expect(result).toMatchSnapshot();
  });
});
