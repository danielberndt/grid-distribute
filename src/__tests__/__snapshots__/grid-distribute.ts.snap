// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`distribute works with many elements 1`] = `
Array [
  Object {
    "element": Object {
      "prio": 4,
    },
    "position": Object {
      "height": 3,
      "left": 2,
      "top": 0,
      "width": 2,
    },
  },
  Object {
    "element": Object {
      "prio": 2,
    },
    "position": Object {
      "height": 2,
      "left": 0,
      "top": 1,
      "width": 1,
    },
  },
  Object {
    "element": Object {
      "prio": 2,
    },
    "position": Object {
      "height": 2,
      "left": 1,
      "top": 1,
      "width": 1,
    },
  },
  Object {
    "element": Object {
      "prio": 1,
    },
    "position": Object {
      "height": 1,
      "left": 0,
      "top": 0,
      "width": 1,
    },
  },
  Object {
    "element": Object {
      "prio": 0.5,
    },
    "position": Object {
      "height": 1,
      "left": 1,
      "top": 0,
      "width": 1,
    },
  },
]
`;

exports[`distribute works with one element 1`] = `
Array [
  Object {
    "element": Object {
      "prio": 2,
    },
    "position": Object {
      "height": 3,
      "left": 0,
      "top": 0,
      "width": 4,
    },
  },
]
`;

exports[`distribute works with two identical elements 1`] = `
Array [
  Object {
    "element": Object {
      "prio": 2,
    },
    "position": Object {
      "height": 3,
      "left": 0,
      "top": 0,
      "width": 2,
    },
  },
  Object {
    "element": Object {
      "prio": 2,
    },
    "position": Object {
      "height": 3,
      "left": 2,
      "top": 0,
      "width": 2,
    },
  },
]
`;
